AUT-678

ri-heme opened the pull request:
Methods for expanding organisms specific databases with side-product metabolites

ri-heme committed:
feat: logging

Add logger module for warnings/debugging.

ri-heme committed:
feat: chemical standardization and utils

Add functions to standardize compounds, calculate molecular fingerprints and similarity between molecules.

ri-heme committed:
feat: reaction rule library

Add reaction rule library, which can be filtered by diameter or compound.

ri-heme committed:
feat: apply rules to compound

Add function to apply available rules to a specificed chemical compound.

ri-heme committed:
feat: accept RDKit objects in functions

Functions may now accept RDKit molecule objects.

ri-heme committed:
feat: store/fetch reaction rules from database

To ease access and manipulation of reaction rules, a SQL database and several functions to fetch rules have been set up.

ri-heme committed:
fix: formatting and docstrings

Added a docstring and took of formatting issues (e.g., consistent use of double quotes).

ri-heme committed:
feat: SQL to select metabolites

Add a function to select metabolites by their InChI depection, to be used for results.

fix: fields used in `apply_to` function

ri-heme committed:
fix: various

fix: add handler to logger
fix: reaction not initialized from SMARTS string
fix: pop filter not using index argument
fix: empty list arguments changed to None
fix: add comments to select functions

ri-heme committed:
feat: list offtargets, context management

Add function to process reaction results into a list of off-target metabolites.

Make rule library a context manager, so database connection is closed on exit.

ri-heme committed:
fix: context manager, logger

Fix logging level and context manager abstract class inheritance.

ri-heme committed:
fix: __all__ attr

As per PEP8, all modules should explicitly declare the names in their public API with the __all__ attribute.

ri-heme committed:
feat: visualization module

Visualization module to plot the results of rule simulations.

ri-heme committed:
feat: open/close db

Add functions to close/open connection to the SQL database.

Rename off-target function.

ri-heme committed:
fix: add UDP as co-factor

ri-heme committed:
fix: in viz, do not override existing edges

When creating the pathway visualization, if an edge already exists, then do not to override it. This prevents it being marked as an off-targeting edge.

ri-heme committed:
fix: organism filter does not require model

Makes the `model_name` parameter of `filter_by_organism` optional.

ri-heme committed:
fix: annotate/BiGGify models

The "small_ecoli" and "iJO1366" model were annotated and their metabolite/reaction IDs were mapped to BiGG IDs.

ri-heme committed:
tests: add tests for pathways module

ri-heme committed:
tests: setup Miniconda

Use Miniconda instead of Python, and install RDKit.

ri-heme committed:
fix: prevent running out of memory

When applying a big number of rules, running out of memory was possible because the simulations were stored in a list. Use a generator instead.

ri-heme committed:
fix: isolate BaseFactory class

`_BaseFactory` is to be shared with other modules, so it deserves its own home.

ri-heme committed:
feat: remote module

The remote module is used to fetch large files we depend on (e.g., reaction rule database).

ri-heme committed:
feat: database downloaded from remote

Add the option to specify a locally-stored reaction rule library. If such database does not exist, then it is downloaded from an external repository.

ri-heme committed:
fix: add/fix __all__

ri-heme committed:
feat: add functions to manipulate databases

Add functions to read, write, and merge struct/mapping files.

ri-heme committed:
fix: import new functions

ri-heme committed:
fix: move new database functions

ri-heme committed:
fix: wrong column reference in merge

ri-heme committed:
fix: split struct/mapping functions

ri-heme committed:
feat: create mapping from struct

ri-heme committed:
feat: additional database creation functions

Functions to create structs from InChI data and rule simulation results.

ri-heme committed:
fix: mapping.from_struct metadata condition

ri-heme committed:
tests: use smaller db for unit tests

ri-heme committed:
fix: return empty df when db is closed

Instead of raising exception, inform user that the database has been closed.

ri-heme committed:
tests: fix filter by compound test

ri-heme committed:
fix: handle empty columns when reading mapping

ri-heme committed:
tests: add unittesting for io module

ri-heme committed:
build: reinstate CI config

CI config was altered during rebasing, so this reinstates it.

ri-heme committed:
build: use conda to run tests

Using conda to run unit tests allows for using libraries like rdkit.

ri-heme committed:
docs: edit documentation of factory variables

ri-heme committed:
build: add version to workflow in config.yml

dmccloskey commented on BFAIR/io/database/mapping.py:
```diff
@@ -0,0 +1,147 @@
+
+from BFAIR.io.database.struct import FILE_COLUMNS as STRUCT_FILE_COLUMNS
+
+FILE_COLUMNS = ["unused_mass", "formula", "ids"]
```
Is there a reason for keeping these as individual functions instead of wrapping them into a class?

ri-heme replied:
Thanks! 🎉

1. I will double check again, though I think all public methods had a test. However, I did not include tests for private methods/modules. For instance, there is a private `_queries` module that the rule library class depends on to create SQL statements. It is implicitly tested by the rule library tests (as most of the class' methods depend on SQL statements).
2. The logger is used by the rule library class to issue info/warning messages under certain conditions (e.g., no rules available after applying a stringent filter).
3. I somehow managed to fix this, so the tests have passed now.
4. I will ping Matthias so he can have a look. 😁

> Is there a reason for keeping these as individual functions instead of wrapping them into a class?

Since the struct/mapping files are handled as dataframes, I thought it made more sense to provide static methods (that return/take dataframes) instead of a container class (that wraps a dataframe).

Example usage:

```
from BFAIR.io import struct, mapping
# reaction rule simulations returning 'results' dict
offtarget_struct = struct.from_products_dict(results)
offtarget_mapping = mapping.from_struct(offtarget_struct)
expanded_mapping = mapping.merge(yeastGEM_mapping, offtarget_mapping)
```

ri-heme committed:
build: rename status checks

dmccloskey commented on BFAIR/io/database/mapping.py:
```diff
@@ -0,0 +1,147 @@
+        input dataframe.
+    """
+    if database_name is None:
+        database_name = _get_metadata(mapping, "database_name")
```
Good checks.  Missing these names will cause lots of its in SmartPeak.

dmccloskey commented on BFAIR/io/remote.py:
```diff
@@ -0,0 +1,118 @@
+                file.write(chunk)
+    size, checksum = get_filesize(destination), get_checksum(destination)
+    if checksum != file_info.expected_checksum or size != file_info.expected_size:
+        raise IOError(f"The downloaded file `{destination}` is corrupted. Remove it and try re-downloading.")
```
Nice check

dmccloskey commented on BFAIR/logger.py:
```diff
@@ -0,0 +1,27 @@
+import logging
+
+
+def get_logger(name):
```
Where is the logger used?

ri-heme replied:
This is used mainly by the rule library class to display warnings or info messages. For instance, if the user applies a filter that reduces the number of rules to zero, a warning will be issued. Another situation: an info level message is also logged when the database is being downloaded for the first time.

dmccloskey replied:
Depending on how the library grows, it might make sense to add in strategic application level logging to help users/developer debug the application.

dmccloskey reviewed (changes requested):
Looks like a lot of hard work went into this!  The documentation and error checking is top notch.

Please double check if you have complete code coverage.  From my count of the files and methods, it looks like some methods do not have tests.

Is the logger used for anything at this stage?

Can you take a look why CircleCI is not triggering?

Can you also invite @mmattano to provide a review as well?  It is a big PR so I probably missed a few things.

ri-heme committed:
build: add missing dependencies

rdkit requires libXrender to be installed in Linux, requests was missing from the requirements.txt

ri-heme committed:
build: add matplotlib to dependencies

ri-heme committed:
test: heed pytest warnings and fix non-passing unit test

mmattano reviewed (commented):
Looks fine to me, really nice work! The code is great, I'll definitely use it as a reference for future modules
